v0.18.0 postinstall missing dependencies

**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
I think some dependencies are missing at the time when `postinstall` scripts are ran.

I'm having this error:
```
yarn install v0.18.0
info No lockfile found.
warning test: No license field
[1/4] Resolving packages...
[2/4] Fetching packages...
warning fsevents@1.0.15: The platform "linux" is incompatible with this module.
info "fsevents@1.0.15" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
[4/4] Building fresh packages...
[1/3] ⠐ gifsicle
[2/3] ⠐ jpegtran-bin
[3/3] ⠐ optipng-bin
[-/3] ⠐ waiting...
error [my-project]/node_modules/jpegtran-bin: Command failed.
Exit code: 1
Command: sh
Arguments: -c node lib/install.js
Directory: [my-project]/node_modules/jpegtran-bin
Output:
module.js:472
    throw err;
    ^

Error: Cannot find module 'ini'
    at Function.Module._resolveFilename (module.js:470:15)
    at Function.Module._load (module.js:418:25)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> ([my-project]/node_modules/rc/lib/utils.js:3:12)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

In another project, after a `yarn upgrade` I also encountered this problem with other packages (`gifsicle`'s `postinstall` missing `deep-extend`). So it seems like a general problem.

**If the current behavior is a bug, please provide the steps to reproduce.**
Given this `package.json`:
```json
{
  "name": "test",
  "devDependencies": {
    "babel-cli": "^6.18.0",
    "gulp-imagemin": "^3.1.1"
  }
}
```
Run `yarn install`.

I noticed that if I try to install only one of these two, it will work. I guess that's an issue with shared dependencies between them.

**What is the expected behavior?**
Installation completes successful.

**Please mention your node.js, yarn and operating system version.**
node.js: v7.2.0
yarn: v0.18.0
OS: Arch Linux

I have same `Error: Cannot find module 'ini'` issue after `yarn install`, but for `global-prefix` dependency. 

Everything if fine on Fedora Desktop, but fails on Ubuntu 16.04.1 server (both system uses same `yarn` and `node` versions).

--- a/__tests__/commands/install/integration.js
+++ b/__tests__/commands/install/integration.js
@@ -696,20 +696,6 @@ test('install a scoped module from authed private registry with a missing traili
   });
 });
 
-test.concurrent('install a module with incompatible optional dependency should skip dependency',
-  (): Promise<void> => {
-    return runInstall({}, 'install-should-skip-incompatible-optional-dep', async (config) => {
-      assert.ok(!(await fs.exists(path.join(config.cwd, 'node_modules', 'dep-incompatible'))));
-    });
-  });
-
-test.concurrent('install a module with incompatible optional dependency should skip transient dependencies',
-  (): Promise<void> => {
-    return runInstall({}, 'install-should-skip-incompatible-optional-dep', async (config) => {
-      assert.ok(!(await fs.exists(path.join(config.cwd, 'node_modules', 'dep-a'))));
-    });
-  });
-
 test.concurrent('install will not overwrite files in symlinked scoped directories', async (): Promise<void> => {
   await runInstall({}, 'install-dont-overwrite-linked-scoped', async (config): Promise<void> => {
     const dependencyPath = path.join(config.cwd, 'node_modules', '@fakescope', 'fake-dependency');
@@ -727,8 +713,22 @@ test.concurrent('install will not overwrite files in symlinked scoped directorie
   });
 });
 
+test.concurrent('install a module with incompatible optional dependency should skip dependency',
+  (): Promise<void> => {
+    return runInstall({}, 'install-should-skip-incompatible-optional-dep', async (config) => {
+      assert.ok(!(await fs.exists(path.join(config.cwd, 'node_modules', 'dep-incompatible'))));
+    });
+  });
+
+test.concurrent('install a module with incompatible optional dependency should skip transient dependencies',
+  (): Promise<void> => {
+    return runInstall({}, 'install-should-skip-incompatible-optional-dep', async (config) => {
+      assert.ok(!(await fs.exists(path.join(config.cwd, 'node_modules', 'dep-a'))));
+    });
+  });
+
 if (process.platform !== 'darwin') {
-  test('install incompatible optional dependency should still install shared child dependencies',
+  test.concurrent('install incompatible optional dependency should still install shared child dependencies',
     (): Promise<void> => {
       // this tests for a problem occuring due to optional dependency incompatible with os, in this case fsevents
       // this would fail on os's incompatible with fsevents, which is everything except osx.
@@ -737,5 +737,21 @@ if (process.platform !== 'darwin') {
         assert.ok(await fs.exists(path.join(config.cwd, 'node_modules', 'ini')));
         assert.ok(await fs.exists(path.join(config.cwd, 'node_modules', 'strip-json-comments')));
       });
+    });
+}
+
+// Covers current behavior, I think this should be changed https://github.com/yarnpkg/yarn/issues/2274
+test.concurrent('optional dependency that fails to build should still be installed',
+  (): Promise<void> => {
+    return runInstall({}, 'should-install-failing-optional-deps', async (config) => {
+      assert.ok(await fs.exists(path.join(config.cwd, 'node_modules', 'optional-failing')));
+    });
+  });
+
+test.concurrent('a subdependency of an optional dependency that fails should be installed',
+  (): Promise<void> => {
+    return runInstall({}, 'should-install-failing-optional-sub-deps', async (config) => {
+      assert.ok(await fs.exists(path.join(config.cwd, 'node_modules', 'optional-failing')));
+      assert.ok(await fs.exists(path.join(config.cwd, 'node_modules', 'sub-dep')));
     });
 }

--- a/__tests__/commands/install/integration.js
+++ b/__tests__/commands/install/integration.js
@@ -740,7 +740,7 @@ if (process.platform !== 'darwin') {
     });
 }
 
-// Covers current behavior, I think this should be changed https://github.com/yarnpkg/yarn/issues/2274
+// Covers current behavior, issue opened whether this should be changed https://github.com/yarnpkg/yarn/issues/2274
 test.concurrent('optional dependency that fails to build should still be installed',
   (): Promise<void> => {
     return runInstall({}, 'should-install-failing-optional-deps', async (config) => {
@@ -754,4 +754,4 @@ test.concurrent('a subdependency of an optional dependency that fails should be 
       assert.ok(await fs.exists(path.join(config.cwd, 'node_modules', 'optional-failing')));
       assert.ok(await fs.exists(path.join(config.cwd, 'node_modules', 'sub-dep')));
     });
-}
+  });

--- a/__tests__/commands/install/integration.js
+++ b/__tests__/commands/install/integration.js
@@ -136,7 +136,7 @@ test.concurrent('hoisting should factor ignored dependencies', async () => {
   });
 });
 
-test.concurrent('--production flag ignores dev dependencies', () => {
+test('--production flag ignores dev dependencies', () => {
   return runInstall({production: true}, 'install-production', async (config) => {
     assert.ok(
       !await fs.exists(path.join(config.cwd, 'node_modules', 'lodash')),
@@ -467,7 +467,7 @@ test.concurrent(
 );
 
 // disabled to resolve https://github.com/yarnpkg/yarn/pull/1210
-test.skip('install should hoist nested bin scripts', (): Promise<void> => {
+test('install should hoist nested bin scripts', (): Promise<void> => {
   return runInstall({binLinks: true}, 'install-nested-bin', async (config) => {
     const binScripts = await fs.walk(path.join(config.cwd, 'node_modules', '.bin'));
     // need to double the amount as windows makes 2 entries for each dependency
@@ -727,13 +727,15 @@ test.concurrent('install will not overwrite files in symlinked scoped directorie
   });
 });
 
-test.concurrent.skip('install incompatible optional dependency should still install shared child dependencies',
-  (): Promise<void> => {
-    // this tests for a problem occuring due to optional dependency incompatible with os, in this case fsevents
-    // this would fail on os's incompatible with fsevents, which is everything except osx.
-    return runInstall({}, 'install-should-not-skip-required-shared-deps', async (config) => {
-      assert.ok(await fs.exists(path.join(config.cwd, 'node_modules', 'deep-extend')));
-      assert.ok(await fs.exists(path.join(config.cwd, 'node_modules', 'ini')));
-      assert.ok(await fs.exists(path.join(config.cwd, 'node_modules', 'strip-json-comments')));
+if (process.platform !== 'darwin') {
+  test('install incompatible optional dependency should still install shared child dependencies',
+    (): Promise<void> => {
+      // this tests for a problem occuring due to optional dependency incompatible with os, in this case fsevents
+      // this would fail on os's incompatible with fsevents, which is everything except osx.
+      return runInstall({}, 'install-should-not-skip-required-shared-deps', async (config) => {
+        assert.ok(await fs.exists(path.join(config.cwd, 'node_modules', 'deep-extend')));
+        assert.ok(await fs.exists(path.join(config.cwd, 'node_modules', 'ini')));
+        assert.ok(await fs.exists(path.join(config.cwd, 'node_modules', 'strip-json-comments')));
+      });
     });
-  });
+}

--- a/__tests__/commands/install/integration.js
+++ b/__tests__/commands/install/integration.js
@@ -467,7 +467,7 @@ test.concurrent(
 );
 
 // disabled to resolve https://github.com/yarnpkg/yarn/pull/1210
-test('install should hoist nested bin scripts', (): Promise<void> => {
+test.skip('install should hoist nested bin scripts', (): Promise<void> => {
   return runInstall({binLinks: true}, 'install-nested-bin', async (config) => {
     const binScripts = await fs.walk(path.join(config.cwd, 'node_modules', '.bin'));
     // need to double the amount as windows makes 2 entries for each dependency

--- a/__tests__/commands/install/integration.js
+++ b/__tests__/commands/install/integration.js
@@ -136,7 +136,7 @@ test.concurrent('hoisting should factor ignored dependencies', async () => {
   });
 });
 
-test('--production flag ignores dev dependencies', () => {
+test.concurrent('--production flag ignores dev dependencies', () => {
   return runInstall({production: true}, 'install-production', async (config) => {
     assert.ok(
       !await fs.exists(path.join(config.cwd, 'node_modules', 'lodash')),

--- a/__tests__/commands/install/integration.js
+++ b/__tests__/commands/install/integration.js
@@ -727,11 +727,11 @@ test.concurrent('install a module with incompatible optional dependency should s
     });
   });
 
+// this tests for a problem occuring due to optional dependency incompatible with os, in this case fsevents
+// this would fail on os's incompatible with fsevents, which is everything except osx.
 if (process.platform !== 'darwin') {
   test.concurrent('install incompatible optional dependency should still install shared child dependencies',
     (): Promise<void> => {
-      // this tests for a problem occuring due to optional dependency incompatible with os, in this case fsevents
-      // this would fail on os's incompatible with fsevents, which is everything except osx.
       return runInstall({}, 'install-should-not-skip-required-shared-deps', async (config) => {
         assert.ok(await fs.exists(path.join(config.cwd, 'node_modules', 'deep-extend')));
         assert.ok(await fs.exists(path.join(config.cwd, 'node_modules', 'ini')));

--- a/__tests__/fixtures/install/should-install-failing-optional-deps/optional-failing/install.js
+++ b/__tests__/fixtures/install/should-install-failing-optional-deps/optional-failing/install.js
@@ -1,0 +1,1 @@
+process.exit(1);

--- a/__tests__/fixtures/install/should-install-failing-optional-deps/optional-failing/package.json
+++ b/__tests__/fixtures/install/should-install-failing-optional-deps/optional-failing/package.json
@@ -1,0 +1,7 @@
+{
+  "name": "optional-failing",
+  "version": "0.0.0",
+  "scripts": {
+    "install": "node install.js"
+  }
+}

--- a/__tests__/fixtures/install/should-install-failing-optional-deps/package.json
+++ b/__tests__/fixtures/install/should-install-failing-optional-deps/package.json
@@ -1,0 +1,5 @@
+{
+  "optionalDependencies": {
+    "optional-failing": "file:optional-failing"
+  }
+}

--- a/__tests__/fixtures/install/should-install-failing-optional-sub-deps/optional-failing/install.js
+++ b/__tests__/fixtures/install/should-install-failing-optional-sub-deps/optional-failing/install.js
@@ -1,0 +1,1 @@
+process.exit(1);

--- a/__tests__/fixtures/install/should-install-failing-optional-sub-deps/optional-failing/package.json
+++ b/__tests__/fixtures/install/should-install-failing-optional-sub-deps/optional-failing/package.json
@@ -1,0 +1,10 @@
+{
+  "name": "optional-failing",
+  "version": "0.0.0",
+  "scripts": {
+    "install": "node install.js"
+  },
+  "dependencies": {
+    "sub-dep": "file:../sub-dep"
+  }
+}

--- a/__tests__/fixtures/install/should-install-failing-optional-sub-deps/package.json
+++ b/__tests__/fixtures/install/should-install-failing-optional-sub-deps/package.json
@@ -1,0 +1,5 @@
+{
+  "optionalDependencies": {
+    "optional-failing": "file:optional-failing"
+  }
+}

--- a/__tests__/fixtures/install/should-install-failing-optional-sub-deps/sub-dep/install.js
+++ b/__tests__/fixtures/install/should-install-failing-optional-sub-deps/sub-dep/install.js
@@ -1,0 +1,1 @@
+process.exit(0);

--- a/__tests__/fixtures/install/should-install-failing-optional-sub-deps/sub-dep/package.json
+++ b/__tests__/fixtures/install/should-install-failing-optional-sub-deps/sub-dep/package.json
@@ -1,0 +1,7 @@
+{
+  "name": "sub-dep",
+  "version": "0.0.0",
+  "scripts": {
+    "install": "node install.js"
+  }
+}

--- a/__tests__/package-hoister.js
+++ b/__tests__/package-hoister.js
@@ -28,7 +28,7 @@ test('Produces valid destination paths for scoped modules', () => {
     _reference: (({}: any): PackageReference),
   }: any): Manifest);
 
-  const info = new HoistManifest(key, parts, pkg, '', false);
+  const info = new HoistManifest(key, parts, pkg, '', () => false);
 
   const tree = new Map([
     ['@scoped/dep', info],

--- a/src/package-hoister.js
+++ b/src/package-hoister.js
@@ -137,7 +137,7 @@ export default class PackageHoister {
         return null;
       }
       // non ignored dependencies inherit parent's ignored status
-      // but parent may transition from ignored to non ignored when hoisting
+      // parent may transition from ignored to non ignored when hoisted if it is used in another non ignored branch
       if (!isIgnored() && parent.isIgnored()) {
         isIgnored = () => !!parent && parent.isIgnored();
       }
@@ -185,7 +185,7 @@ export default class PackageHoister {
       const existing = this.tree.get(checkKey);
       if (existing) {
         if (existing.loc === info.loc) {
-          // deduping an unignored reference to an ignored one
+          // switch to non ignored if earlier deduped version was ignored
           if (existing.isIgnored() && !info.isIgnored()) {
             existing.isIgnored = info.isIgnored;
           }

--- a/src/package-hoister.js
+++ b/src/package-hoister.js
@@ -130,7 +130,7 @@ export default class PackageHoister {
 
     //
     let parentParts: Parts = [];
-    let isIgnored = () => { return ref.ignore; };
+    let isIgnored = () => ref.ignore;
 
     if (parent) {
       if (!this.tree.get(parent.key)) {
@@ -139,7 +139,7 @@ export default class PackageHoister {
       // non ignored dependencies inherit parent's ignored status
       // but parent may transition from ignored to non ignored when hoisting
       if (!isIgnored() && parent.isIgnored()) {
-        isIgnored = () => { return !!parent && parent.isIgnored(); };
+        isIgnored = () => !!parent && parent.isIgnored();
       }
       parentParts = parent.parts;
     }

--- a/src/package-hoister.js
+++ b/src/package-hoister.js
@@ -13,8 +13,8 @@ type Parts = Array<string>;
 let historyCounter = 0;
 
 export class HoistManifest {
-  constructor(key: string, parts: Parts, pkg: Manifest, loc: string, isIgnored: boolean) {
-    this.ignore = isIgnored;
+  constructor(key: string, parts: Parts, pkg: Manifest, loc: string, isIgnored: () => boolean) {
+    this.isIgnored = isIgnored;
     this.loc = loc;
     this.pkg = pkg;
 
@@ -27,7 +27,7 @@ export class HoistManifest {
     this.addHistory(`Start position = ${key}`);
   }
 
-  ignore: boolean;
+  isIgnored: () => boolean;
   pkg: Manifest;
   loc: string;
   parts: Parts;
@@ -130,13 +130,17 @@ export default class PackageHoister {
 
     //
     let parentParts: Parts = [];
-    let isIgnored = ref.ignore;
+    let isIgnored = () => { return ref.ignore; };
 
     if (parent) {
       if (!this.tree.get(parent.key)) {
         return null;
       }
-      isIgnored = isIgnored || parent.ignore;
+      // non ignored dependencies inherit parent's ignored status
+      // but parent may transition from ignored to non ignored when hoisting
+      if (!isIgnored() && parent.isIgnored()) {
+        isIgnored = () => { return parent.isIgnored(); };
+      }
       parentParts = parent.parts;
     }
 
@@ -182,8 +186,8 @@ export default class PackageHoister {
       if (existing) {
         if (existing.loc === info.loc) {
           // deduping an unignored reference to an ignored one
-          if (existing.ignore && !info.ignore) {
-            existing.ignore = false;
+          if (existing.isIgnored() && !info.isIgnored()) {
+            existing.isIgnored = info.isIgnored;
           }
 
           existing.addHistory(`Deduped ${fullKey} to this item`);
@@ -277,7 +281,6 @@ export default class PackageHoister {
     // remove this item from the `tree` map so we can ignore it
     this.tree.delete(key);
 
-    //
     const {parts, duplicate} = this.getNewParts(key, info, rawParts.slice());
     const newKey = this.implodeKey(parts);
     const oldKey = key;
@@ -389,7 +392,7 @@ export default class PackageHoister {
       const ref = info.pkg._reference;
       invariant(ref, 'expected reference');
 
-      if (info.ignore) {
+      if (info.isIgnored()) {
         info.addHistory('Deleted as this module was ignored');
       } else {
         visibleFlatTree.push([loc, info]);

--- a/src/package-hoister.js
+++ b/src/package-hoister.js
@@ -139,7 +139,7 @@ export default class PackageHoister {
       // non ignored dependencies inherit parent's ignored status
       // but parent may transition from ignored to non ignored when hoisting
       if (!isIgnored() && parent.isIgnored()) {
-        isIgnored = () => { return parent.isIgnored(); };
+        isIgnored = () => { return !!parent && parent.isIgnored(); };
       }
       parentParts = parent.parts;
     }

